src/platforms/stm32/src/lib/sys.c

```diff
@@ -148,19 +153,74 @@ Module *sys_load_module(GlobalContext *global, const char *module_name)
 
 Context *sys_create_port(GlobalContext *glb, const char *driver_name, term opts)
 {
-    Context *new_ctx = context_new(glb);
-
-    if (!strcmp(driver_name, "gpio")) {
-        gpiodriver_init(new_ctx);
-    } else {
-        context_destroy(new_ctx);
-        return NULL;
+    Context *new_ctx = port_driver_create_port(driver_name, glb, opts);
+    if (IS_NULL_PTR(new_ctx)) {
+        AVM_LOGE(TAG, "Failed to load port \"%s\".  Ensure the port is configured properly in the build.", driver_name);
+        new_ctx = NULL;
     }
-
     return new_ctx;
 }
 
 term sys_get_info(Context *ctx, term key)
 {
     return UNDEFINED_ATOM;
 }
+
+void port_driver_init_all(GlobalContext *global)
+{
+    for (struct PortDriverDefListItem *item = port_driver_list; item != NULL; item = item->next) {
```

Comment:
Should these for loops be using FOREACH macros out of lists.h, or am I missing something about how these structs are structured?

Reply:
That is a very good question. I implemented this the same way it was done on the esp32, I just copied this function directly from the esp32 port `sys.c` expecting to need to make corrections, but when I tested it with my rewrite of the stm32 `gpio_driver` it seemed to work fine. I will investigate this a little closer to see if I missed something here.

Reply:
This structs are not based on list.h, so you cannot use them. Here last item makes use of NULL differently from `list.h`.

Reply:
As far as I can tell the `port_driver_list` is always empty when this is called, and its primary purpose is to store the GlobalContext *ptr for the `port_driver_init_cb`, to be used later when drivers will be added to the `port_driver_list` when `port_driver_create_port` is used to create a port for a driver.